### ISSUE TITLE
performance of ColumnParallelLinear and RowParallelLinear

### ISSUE THREAD

**Issue by yingtongxiong** [HazyResearch/flash-attention#201]
Hello, I have tested the performance of columnparallelLinear and rowParallelLinear recently. However, I found some problems. The first one is that the gradients of these two layers are different from torch.nn.Linear, which has quiet large gap. The second is that the speed of these two layers are slower than torch.nn.Linear, which is very confused to me. Can you have any suggestions?
The following is my test code:

`

def test_Row():
    
    init()
    colossalai.launch_from_torch(config={
        'parallel': {
            'tensor': {
                'size': 2,  # 初始化一个tp size为2的
                'mode': '1d'
            }
        }, 
    }, seed=42)
    
    world_size = int(os.environ["WORLD_SIZE"])
    parallel_state.initialize_model_parallel(tensor_model_parallel_size_=world_size)
    rank = parallel_state.get_tensor_model_parallel_rank()
    process_group=gpc.get_group(ParallelMode.PARALLEL_1D)
    
    x_pt = torch.randn(batch_size * seqlen, in_features, device=device, dtype=dtype,requires_grad=True)
    
    x = tensor_parallel.scatter_to_tensor_model_parallel_region(x_pt).detach().clone().requires_grad_()
    
    model_pt = torch.nn.Linear(in_features, out_features, bias=bias, device=device, dtype=dtype)

    partition_in_features = in_features // world_size

    model = RowParallelLinear(in_features, out_features,
                              process_group, bias=bias,
                              sequence_parallel=sequence_parallel, device=device, dtype=dtype)
    
    with torch.no_grad():
        model.weight.copy_(
            model_pt.weight[:, rank * partition_in_features:(rank + 1) * partition_in_features]
        )
        if bias and rank == 0:
            model.bias.copy_(
                model_pt.bias)

    out = model(x)
    out_pt = model_pt(x_pt)
    
    loss_pt = torch.rand_like(out_pt) / 32
    
    partition_in = batch_size * seqlen // world_size
   
    if sequence_parallel:
        loss = loss_pt[rank * partition_in:(rank + 1) * partition_in, :]
        out.backward(loss)
        out_pt.backward(loss_pt)
        assert torch.allclose(
            out, out_pt[rank * partition_in:(rank + 1) * partition_in, :], rtol=rtol, atol=atol)
        assert torch.allclose(model.weight.grad, model_pt.weight.grad[:, rank * partition_in_features:(rank + 1) * partition_in_features], atol=3e-2, rtol=1e-2)
    else:
        loss = loss_pt
        assert torch.allclose(out, out_pt, rtol=rtol, atol=atol)
        out.backward(loss)
        out_pt.backward(loss_pt)
        assert torch.allclose(model.weight.grad, model_pt.weight.grad[:, rank * partition_in_features:(rank + 1) * partition_in_features], atol=atol*10, rtol=rtol)
        
    
    benchmark_forward(model, x, repeats=repeats, desc='Flash Parallel Linear')
    benchmark_forward(model_pt, x_pt, repeats=repeats, desc='Pytorch Linear')
    
    benchmark_backward(model, x, repeats=repeats, desc='Flash Parallel Linear')
    benchmark_backward(model_pt, x_pt, repeats=repeats, desc='Pytorch Linear')
    
    benchmark_combined(model, x, repeats=repeats, desc='Flash Parallel Linear')
    benchmark_combined(model_pt, x_pt, repeats=repeats, desc='Pytorch Linear')

def test_Column():
    
    init()
    colossalai.launch_from_torch(config={
        'parallel': {
            'tensor': {
                'size': 2,  # 初始化一个tp size为2的
                'mode': '1d'
            }
        }, 
    }, seed=42)
    world_size = int(os.environ["WORLD_SIZE"])
    
    parallel_state.initialize_model_parallel(tensor_model_parallel_size_=world_size)
    rank = gpc.get_global_rank()
    process_group=gpc.get_group(ParallelMode.PARALLEL_1D)
    input_pt = torch.randn(batch_size * seqlen, in_features, device=device, dtype=dtype, requires_grad=True)
    
    if sequence_parallel:
        input = tensor_parallel.scatter_to_sequence_parallel_region(input_pt).detach().clone().requires_grad_()
    else:
        input = input_pt.detach().clone().requires_grad_()
        
    model_pt = torch.nn.Linear(in_features, out_features, bias=bias, device=device, dtype=dtype)
    model = ColumnParallelLinear(in_features, out_features, process_group=process_group, 
                                 sequence_parallel=sequence_parallel, bias=bias, device=device, dtype=dtype)

    
    partition_out_features = out_features // world_size
    
    
    with torch.no_grad():
        model.weight.copy_(
            model_pt.weight[rank * partition_out_features:(rank + 1) * partition_out_features, :]
        )
        if bias:
            model.bias.copy_(
                model_pt.bias[rank * partition_out_features:(rank + 1) * partition_out_features]
            )
    
    output = model(input)
    output_pt = model_pt(input_pt)
    
    loss_pt = torch.rand_like(output_pt) / 32
    
    assert torch.allclose(output, output_pt[:, rank * partition_out_features:(rank + 1) * partition_out_features])
    loss = loss_pt[:,  rank * partition_out_features:(rank + 1) * partition_out_features]
  
    output.backward(loss)
    output_pt.backward(loss_pt)
    assert torch.allclose(model.weight.grad, model_pt.weight.grad[rank * partition_out_features:(rank + 1) * partition_out_features, :], atol=3e-2, rtol=1e-2)
    assert torch.allclose(model.bias.grad, model_pt.bias.grad[rank * partition_out_features:(rank + 1) * partition_out_features], atol=3e-2, rtol=1e-2)
    
    benchmark_forward(model_pt, input_pt, repeats=repeats, desc="Pytorch Linear")
    benchmark_forward(model, input, repeats=repeats, desc="Parallel Linear")
`
The benchmark functions are from your flash-attn's benchmark.

**Comment by tridao** [comment]
They're meant to be reimplementation of Megatron's [ColumnParallelLinear](https://github.com/NVIDIA/Megatron-LM/blob/3db2063b1ff992a971ba18f7101eecc9c4e90f03/megatron/core/tensor_parallel/layers.py#L418) and RowParallelLinear.
You should compare the error (this implementation - nn.Linear) vs (Megatron implementation - nn.Linear).

As for speed, depending on sizes, ColumnParallelLinear and RowParallleLinear could be slower than nn.Linear since there's communication between GPUs.
